### ISSUE TITLE
Another subtle bug

### ISSUE THREAD

**Issue by lukaszsamson** [anvaka/set-vs-object#2]
The object counter will not handle strings matching special JS properties, eg. '**proto**'.

**Comment by anvaka** [comment]
This is exactly what I had in mind :). Though `proto` still works correctly here:

``` js
function getUniqueElements(array) {
  var counter = {};
  for (var i = 0; i < array.length; ++i) {
    var key = array[i];
    counter[key] = 1;
  }
  return Object.keys(counter);
}

console.log(getUniqueElements(['proto']));
```

Do you know how to break it? And after that, do you know how to fix it?

**Comment by erykpiast** [comment]
Passing string `'__proto__'` will break the code, `Object.keys(counter)` will  return empty array.

**Comment by lukaszsamson** [comment]
Yes erykpiast, I obviously meant `'__proto__'`, just forgot to escape underscores properly.

**Comment by erykpiast** [comment]
:+1:

**Comment by anvaka** [comment]
:+1: @erykpiast !

How would recommend to fix it?

**Comment by erykpiast** [comment]
Use real `Set` instead of this poor-man alternative that hashmap is :) Or just approach similiar to [`lodash` one](https://github.com/lodash/lodash/blob/master/lodash.js#L2903-L2944), without using `Object.keys()`. Ugly `if` for `__proto__` name is not the way.

**Comment by anvaka** [comment]
True story: I had this issue in http://anvaka.github.io/pm/#/ - it was very fun to debug.

If you absolutely have to use ES5 (which doesn't have Sets) you can overcome it by using `Object.create(null)`:

``` js
function getUniqueElements(array) {
  var counter = Object.create(null);
  for (var i = 0; i < array.length; ++i) {
    var key = array[i];
    counter[key] = 1;
  }
  return Object.keys(counter);
}

console.log(getUniqueElements(['__proto__']));
```

**Comment by erykpiast** [comment]
Hah, this is smart!
